File: benchmarks/medium/M3-hashmap-stress/javascript/main.js
function main() {
    const N = 5_000_000;
    const LOOKUPS = 1_000_000;
    const RUNS = 10;

    const insertTimings = [];
    const lookupTimings = [];

    console.log("=== BENCHMARK: M3 - Hash Map Stress Test ===");
    console.log("Language: JavaScript");
    console.log(`Runtime:  Node.js ${process.version}`);
    console.log();

    let verifySum = 0n;

    for (let run = 1; run <= RUNS; run++) {
        // Insert phase
        const startInsert = process.hrtime.bigint();
        const map = new Map();
        for (let i = 0; i < N; i++) {
            map.set(`user_${i}`, BigInt(i) * 31n + 7n);
        }
        const insertElapsed = Number(process.hrtime.bigint() - startInsert) / 1e6;
        insertTimings.push(insertElapsed);

        // Lookup phase
        const startLookup = process.hrtime.bigint();
        let state = 42;
        let sum = 0n;
        for (let j = 0; j < LOOKUPS; j++) {
            state = (Math.imul(state, 1664525) + 1013904223) >>> 0;
            const idx = state % N;
            sum += map.get(`user_${idx}`);
        }
        const lookupElapsed = Number(process.hrtime.bigint() - startLookup) / 1e6;
        lookupTimings.push(lookupElapsed);

        verifySum = sum;
        console.log(`Run ${String(run).padStart(2)}: Insert: ${insertElapsed.toFixed(2).padStart(8)} ms | Lookup: ${lookupElapsed.toFixed(2).padStart(8)} ms | Total: ${(insertElapsed + lookupElapsed).toFixed(2).padStart(8)} ms`);
    }

    console.log();
    printStats("Insert", insertTimings);
    printStats("Lookup", lookupTimings);

    const totalTimings = insertTimings.map((v, i) => v + lookupTimings[i]);
    printStats("Total", totalTimings);

    console.log(`Verification: sum = ${verifySum}`);
}

function printStats(label, timings) {
    const min = Math.min(...timings);
    const max = Math.max(...timings);
    const avg = timings.reduce((a, b) => a + b, 0) / timings.length;
    const variance = timings.reduce((a, t) => a + (t - avg) ** 2, 0) / timings.length;
    const stddev = Math.sqrt(variance);

    console.log(`${label}:`);
    console.log(`  Min:    ${min.toFixed(2)} ms`);
    console.log(`  Avg:    ${avg.toFixed(2)} ms`);
    console.log(`  Max:    ${max.toFixed(2)} ms`);
    console.log(`  StdDev: ${stddev.toFixed(2)} ms`);
    console.log();
}

main();

File: benchmarks/medium/M3-hashmap-stress/rust/main.rs
use std::collections::HashMap;
use std::time::Instant;

const N: usize = 5_000_000;
const LOOKUPS: usize = 1_000_000;
const RUNS: usize = 10;

fn main() {
    let mut insert_timings = [0.0f64; RUNS];
    let mut lookup_timings = [0.0f64; RUNS];

    println!("=== BENCHMARK: M3 - Hash Map Stress Test ===");
    println!("Language: Rust");
    println!("Runtime:  rustc (native)");
    println!();

    let mut verify_sum: i64 = 0;

    for run in 0..RUNS {
        // Insert phase
        let start = Instant::now();
        let mut map: HashMap<String, i64> = HashMap::with_capacity(N);
        for i in 0..N {
            let key = format!("user_{}", i);
            let value = (i as i64) * 31 + 7;
            map.insert(key, value);
        }
        let insert_elapsed = start.elapsed().as_secs_f64() * 1000.0;
        insert_timings[run] = insert_elapsed;

        // Lookup phase
        let start = Instant::now();
        let mut state: u32 = 42;
        let mut sum: i64 = 0;
        for _ in 0..LOOKUPS {
            state = state.wrapping_mul(1664525).wrapping_add(1013904223);
            let idx = (state as u64 % N as u64) as usize;
            let key = format!("user_{}", idx);
            sum += map[&key];
        }
        let lookup_elapsed = start.elapsed().as_secs_f64() * 1000.0;
        lookup_timings[run] = lookup_elapsed;

        verify_sum = sum;
        println!(
            "Run {:2}: Insert: {:>8.2} ms | Lookup: {:>8.2} ms | Total: {:>8.2} ms",
            run + 1,
            insert_elapsed,
            lookup_elapsed,
            insert_elapsed + lookup_elapsed
        );
    }

    println!();

    // Insert stats
    let min = insert_timings.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = insert_timings.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let avg = insert_timings.iter().sum::<f64>() / RUNS as f64;
    let variance = insert_timings.iter().map(|t| (t - avg).powi(2)).sum::<f64>() / RUNS as f64;
    let stddev = variance.sqrt();
    println!("Insert:");
    println!("  Min:    {:.2} ms", min);
    println!("  Avg:    {:.2} ms", avg);
    println!("  Max:    {:.2} ms", max);
    println!("  StdDev: {:.2} ms", stddev);
    println!();

    // Lookup stats
    let min = lookup_timings.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = lookup_timings.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let avg = lookup_timings.iter().sum::<f64>() / RUNS as f64;
    let variance = lookup_timings.iter().map(|t| (t - avg).powi(2)).sum::<f64>() / RUNS as f64;
    let stddev = variance.sqrt();
    println!("Lookup:");
    println!("  Min:    {:.2} ms", min);
    println!("  Avg:    {:.2} ms", avg);
    println!("  Max:    {:.2} ms", max);
    println!("  StdDev: {:.2} ms", stddev);
    println!();

    // Total stats
    let total_timings: Vec<f64> = insert_timings
        .iter()
        .zip(lookup_timings.iter())
        .map(|(i, l)| i + l)
        .collect();
    let min = total_timings.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = total_timings.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let avg = total_timings.iter().sum::<f64>() / RUNS as f64;
    let variance = total_timings.iter().map(|t| (t - avg).powi(2)).sum::<f64>() / RUNS as f64;
    let stddev = variance.sqrt();
    println!("Total:");
    println!("  Min:    {:.2} ms", min);
    println!("  Avg:    {:.2} ms", avg);
    println!("  Max:    {:.2} ms", max);
    println!("  StdDev: {:.2} ms", stddev);
    println!();

    println!("Verification: sum = {}", verify_sum);
}
